Feature aws s3 upload

clothe09986 opened the pull request.

clothe09986 committed:
feat: s3 multipart upload

clothe09986 committed:
feat: change multipart upload bucket & insert attachment

clothe09986 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [【影片】媒體庫、前台串接aws影片](https://trello.com/c/wXRAVpZT/4477-%E3%80%90%E5%BD%B1%E7%89%87%E3%80%91%E5%AA%92%E9%AB%94%E5%BA%AB%E3%80%81%E5%89%8D%E5%8F%B0%E4%B8%B2%E6%8E%A5aws%E5%BD%B1%E7%89%87)

jackey8616 commented on src/utility/storage/storage.controller.ts:
```diff
@@ -2,21 +2,76 @@ import { Body, Controller, Post } from '@nestjs/common';
+  @Post('/multipart/complete')
+  async completeMultipartUpload(
+    @Body()
+    body: {
+      params: {
+        Key: string;
+        UploadId: string;
+        MultipartUpload: CompletedMultipartUpload;
+      };
+      file: {
+        name: string;
+        type: string;
+        size: number;
+      };
+      appId: string;
+      authorId: string;
+      attachmentId: string;
+    },
```
Shall write in DTO.

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.service.ts:
```diff
@@ -81,6 +94,40 @@ export class StorageService {
+    return getSignedUrl(this.s3(this.awsS3RegionStorage), command, { expiresIn });
+  }
+
+  async insertVideoAttachment(
+    appId: string,
+    authorId: string,
+    attachmentId: string,
+    file: { name: string; type: string; size: number },
+  ) {
+    const AttachmentRepo = this.entityManager.getRepository(Attachment);
+    return await AttachmentRepo.insert({
+      id: attachmentId,
+      appId,
+      author: { id: authorId },
+      name: file.name,
+      filename: file.name,
+      contentType: file.type,
+      size: file.size,
+      options: [{ source: 's3' }],
+    });
+  }
```
Method `insertVideoAttachment` shall write in MediaInfra & wrap in @ MediaService.  
MediaService should import @ `StorageController`'s constructor and invoke @ there.

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.controller.ts:
```diff
@@ -2,21 +2,76 @@ import { Body, Controller, Post } from '@nestjs/common';
+  ) {
+    const { params, file, appId, authorId, attachmentId } = body;
+    const result = await this.storageService.completeMultipartUpload(params);
+    await this.storageService.insertVideoAttachment(appId, authorId, attachmentId, file);
```
`insertVideoAttachment ` should not write in `storageService`, it should be `mediaService`.

clothe09986 replied:
fixed

clothe09986 committed:
fix: move insert attachment to MediaService

clothe09986 committed:
feat: cloudfront signed cookies

jackey8616 commented on src/media/media.service.ts:
```diff
@@ -0,0 +1,58 @@
+    @InjectEntityManager() private readonly entityManager: EntityManager,
+    private readonly mediaInfra: MediaInfrastructure,
+    private readonly configService: ConfigService<{
+      AWS_CLOUDFRONT_KEY_PAIR_ID: string;
+      AWS_CLOUDFRONT_PRIVATE_KEY: string;
```
These keys is not add into `.env` & `.env.example`  .

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.controller.ts:
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
+    const { params, file, appId, authorId, attachmentId } = body;
+    const result = await this.storageService.completeMultipartUpload(params);
+    await this.mediaService.insertAttachment(appId, authorId, attachmentId, file.name, file.type, file.size, {
+      source: 's3',
```
```ts
{
  source: 's3,
}
```

or 

```ts
{
  source: {
    s3: 's3://.....',
  },
}
```
?

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.service.ts:
```diff
@@ -10,25 +10,34 @@ import {
   constructor(
     private readonly configService: ConfigService<{
       AWS_S3_BUCKET_STATIC: string;
+      AWS_S3_REGION_STATIC: string;
```
This key is not add into `.env` & `.env.example`  .

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.controller.ts:
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
+    },
+  ) {
+    const { params } = body;
+    const { UploadId } = await this.storageService.createMultipartUpload(params);
```
Exposing too many parameters.  
Please wrap `createMultipartUpload` method with limited parameters.

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.controller.ts:
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
+    },
+  ) {
+    const { params } = body;
+    const presignedUrl = await this.storageService.getSignedUrlForUploadPartStorage(params, 60);
```
Exposing too many parameters.  
Please wrap `getSignedUrlForUploadPartStorage ` method with limited parameters.

clothe09986 replied:
fixed

jackey8616 commented on src/utility/storage/storage.controller.ts:
```diff
@@ -2,21 +2,79 @@ import { Body, Controller, Post } from '@nestjs/common';
+    },
+  ) {
+    const { params, file, appId, authorId, attachmentId } = body;
+    const result = await this.storageService.completeMultipartUpload(params);
```
Exposing too many parameters.  
Please wrap `completeMultipartUpload ` method with limited parameters.

clothe09986 replied:
fixed

jackey8616 commented on src/media/media.controller.ts:
```diff
@@ -0,0 +1,19 @@
+export class MediaController {
+  constructor(private readonly mediaService: MediaService) {}
+
+  @Post('/sign-cookies') signCookie(@Res() res: Response, @Body() body: { url: string }) {
+    const cookies = this.mediaService.signCookies(body.url);
+    res.cookie('CloudFront-Expires', cookies['CloudFront-Expires']);
+    res.cookie('CloudFront-Key-Pair-Id', cookies['CloudFront-Key-Pair-Id']);
+    res.cookie('CloudFront-Signature', cookies['CloudFront-Signature']);
+    res.send('success set cookie');
+  }
```
Should `signedCookie` be inside `AuthService` & `AuthController` ?

clothe09986 replied:
or `signedCookie` method is inside `AuthService` and `MediaController` invokes this method?

clothe09986 replied:
fixed

clothe09986 committed:
feat: add env

clothe09986 committed:
fix: remove log

clothe09986 committed:
feat: modify storage service parameters and add DTO

clothe09986 committed:
fix: move sign cloudfront cookies to AuthService

aniku777tw committed:
fix: completeMultipartUpload insert option

aniku777tw committed:
feat: sign cookie controller

aniku777tw committed:
feat: sign cloudfront cookie

clothe09986 committed:
feat: add attachment duration

aniku777tw committed:
refactor: upsert attachment

aniku777tw committed:
feat: put s3 storage location

clothe09986 committed:
fix: attahcment options no s3 source

aniku777tw committed:
fix: sign s3 put command

aniku777tw committed:
fix: auth when sign cookie and sign url

aniku777tw committed:
feat: upload captions

aniku777tw committed:
feat:  upload and get captions

aniku777tw committed:
refactor: manifest parse

clothe09986 committed:
feat: upsert attachment updated_at

aniku777tw committed:
fix: sort captions and remove m3u8 subtitle

aniku777tw committed:
refactor: move sign url to video controller

aniku777tw committed:
refactor: move sign url to video controller

clothe09986 committed:
Refactor attachment name handling in
StorageController

This change modifies the StorageController to use the `attachmentName`
parameter instead of the `name` parameter for setting the name and
filename of the attachment. The `name` field of the file parameter is
now assigned to `fileName`. This refactoring provides more clarity and
consistency in the code.

Fixes #123

aniku777tw committed:
fix: check auth token when get m3u8
